File: minesweeper/java-minesweeper/src/main/java/it/chicio/minesweeper/field/formatter/FieldFormatterWithNewlines.java
package it.chicio.minesweeper.field.formatter;

import it.chicio.minesweeper.field.Field;

public class FieldFormatterWithNewlines implements FieldFormatter {
    public String format(Field field) {
        StringBuilder fieldAsString = new StringBuilder();
        for (int row = 0; row < field.numberOfRows(); row++) {
            for (int column = 0; column < field.numberOfColumn(); column++) {
                fieldAsString.append(field.get(row, column));
                appendSpaceIfIsNotLastColumn(field, fieldAsString, column);
            }
            fieldAsString.append(System.getProperty("line.separator"));
        }
        return fieldAsString.toString();
    }

    private void appendSpaceIfIsNotLastColumn(Field field, StringBuilder fieldAsString, int column) {
        if (isNotLastColumn(field, column)) {
            fieldAsString.append(" ");
        }
    }

    private boolean isNotLastColumn(Field field, int column) {
        return column != field.numberOfColumn() - 1;
    }
}

File: minesweeper/kotlin-mockk-minesweeper/src/main/kotlin/it/chicio/minesweeper/field/formatter/FieldFormatterWithNewlines.kt
package it.chicio.minesweeper.field.formatter

import it.chicio.minesweeper.field.Field

class FieldFormatterWithNewlines : FieldFormatter {
    override fun format(field: Field): String {
        val fieldAsString = StringBuilder()
        for (row in 0 until field.numberOfRows) {
            for (column in 0 until field.numberOfColumn) {
                fieldAsString.append(field[row, column])
                appendSpaceIfIsNotLastColumn(field, fieldAsString, column)
            }
            fieldAsString.append(System.getProperty("line.separator"))
        }
        return fieldAsString.toString()
    }

    private fun appendSpaceIfIsNotLastColumn(field: Field?, fieldAsString: StringBuilder, column: Int) {
        if (isNotLastColumn(field, column)) {
            fieldAsString.append(" ")
        }
    }

    private fun isNotLastColumn(field: Field?, column: Int): Boolean = column != (field?.numberOfColumn ?: 0) - 1
}